### PR TITLE
Allow dispatch region formation to fuse producer and consumer when producer has multiple uses.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Current heuristic to create dispatch regions has checks to fuse
producer with consumer only when the producer has a single use. This
is pessimizing. The only real requirement is that the consumer
dominates all the producers users. Use dominance info to relax this
requirement.

All the models we track currently though dont seem to benefit with
this change (they dont degrade either). The MobileBertInt8 model has
number of dispatches reduce from  1520 -> 1496.

### CODE REVIEW

**Review by antiagainst — APPROVED** [review]
Nice!

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 1]
```diff
@@ -897,17 +898,31 @@ static bool isFusableWithConsumer(OpOperand &use) {
   return areFusableLinalgOps(use);
 }
 
+/// For all uses of an operation, finds the use that dominates all other uses.
+static Optional<OpOperand *> getFusableUse(Operation *op,
+                                           DominanceInfo &dominanceInfo) {
+  for (auto &use : op->getUses()) {
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:904]: I wonder if we can simply use getUsers()?

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 2]
```diff
@@ -897,17 +898,31 @@ static bool isFusableWithConsumer(OpOperand &use) {
   return areFusableLinalgOps(use);
 }
 
+/// For all uses of an operation, finds the use that dominates all other uses.
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:901]: nit: finds -> find

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 3]
```diff
@@ -940,7 +958,8 @@ static void fuseRootsWithConsumers(MLIRContext *context,
 /// fuse with multiple root operations (i.e. replicated). For now a very simple
 /// heuristic is used below, but the mechanism should be general enough to
 /// capture any heuristic.
-static unsigned decideFusableLinalgOps(FunctionOpInterface funcOp) {
+static unsigned decideFusableLinalgOps(FunctionOpInterface funcOp,
+                                       DominanceInfo &dominanceInfo) {
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:962]: use 'const'.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 4]
```diff
@@ -897,17 +898,31 @@ static bool isFusableWithConsumer(OpOperand &use) {
...
+  }
+  return llvm::None;
+}
+
 /// Fuses roots with its consumers. If a root is fused with its consumer, it is
 /// no more tagged as a root to aid with the dispatch region formation.
 static void fuseRootsWithConsumers(MLIRContext *context,
-                                   ArrayRef<Operation *> roots) {
+                                   ArrayRef<Operation *> roots,
+                                   DominanceInfo &dominanceInfo) {
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:919]: use 'const'.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 5]
```diff
@@ -897,17 +898,31 @@ static bool isFusableWithConsumer(OpOperand &use) {
   return areFusableLinalgOps(use);
 }
 
+/// For all uses of an operation, finds the use that dominates all other uses.
+static Optional<OpOperand *> getFusableUse(Operation *op,
+                                           DominanceInfo &dominanceInfo) {
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:903]: use 'const'.

**Review by okkwon — APPROVED** [review]
Added some minor comments. LGTM! Thanks.

**Comment by MaheshRavishankar** [comment]
Thanks for the reviews! It seems like this is hitting some issues downstream. Couple of ways to fix it that I am going to take a look at once I get back after break (in a week). So not going to land this for now.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp]: The method returns the `use`. So iterating over uses makes more sense.

**Pull request merged**